src-ext/TypeScript/ggvega/src/itmEncodingObject.ts

```diff
@@ -0,0 +1,219 @@
+  return itmEncodingObject;
+}
+
+/**
+ * Modify an intermediate `encoding` object using an `AesParams` object
+ *
+ * @remarks
+ * This function will have side-effects, as the argument `itmEncodingObject`
+ * is mutable, and is modified inside this function. We return this object
+ * to signify that we intend to modify it, but we suspect that this is not
+ * strictly necessary; the function could return `null` to signify that it
+ * is called for side-effects. I am curious to know the convention here.
+ *
+ * This function modifies an intermdiate `encoding` object using a
+ * ggspec `AesParams` object. In the future, we may need the `geom`.
+ *
+ * A ggspec `AesParams` element is a key-value pair, where the key
+ * is a ggplot2 aesthetic-name, and the value can be a `string`, `number`,
+ * or `boolean` (?).
+ *
+ * For each `AesParams` element, an `encoding` object is created, and put
+ * into `itmEncodingObject` using the ggplot2 aesthetic-name as a key. The
+ * value (of this key-value pair) is an object with a single key-value
+ * pair: the key is `value`; its value is the value in "visual" space.
+ *
+ * There are helper functions to translate values for shape, color, and size.
+ *
+ * In the future, we may be able to take into account values that are specified
+ * using "data" space: <https://github.com/vega/vega-lite/issues/1601>
+ *
+ * **Called by**
+ * @see {@link itmLayer} to create an intermediate layer
+ *
+ * **Calls**
+ * @see {@link encodingByAes} to create an empty encoding
+ * @see {@link encodingValueShape} to translate shape values
+ * @see {@link encodingValueColor} to translate color values
+ * @see {@link encodingValueSize} to translate size values
+ *
+ *
+ * @param itmEncodingObject
+ * @param ggAesParamsObject
+ *
+ * @returns `ItmEncodingObject`
+ */
+export function itmEncodingObjectByAesParamsObject(
+  itmEncodingObject: ItmEncodingObject,
+  ggAesParamsObject: GG.AesParams
+): ItmEncodingObject {
+  // NOTE: we may have to pass the `geom` from itmLayer()
+
+  // for each member of gsAesParams:
+  //   - extract information from aesParams
+  //   - create ItmEncoding
+  //   - populate ItmEncoding
+  //   - put ItmEncoding into itmEncodingObject
+  for (const aesName in ggAesParamsObject) {
+    if (Object.prototype.hasOwnProperty.call(ggAesParamsObject, aesName)) {
+      // extract information from aes_params
+
+      // NOTE @wenyu: Maybe the `value` can have other types
+      let value: string | number | undefined = ggAesParamsObject[aesName as keyof GG.AesParams];
+
+      /**
+       * keep in mind that values are interpreted
+       * in the "visual" space, not the "data" space
+       *
+       * this will be an issue for geom_hline and geom_vline,
+       * as their x & y aes_params are specified in "data" space
+       *
+       * issue: https://github.com/vega/vega-lite/issues/1601
+       * PR: https://github.com/vega/vega-lite/pull/4201
+       */
+
+      // tranlsate
+      if (aesName == 'shape') {
```

Comment:
Instead of using `if/else`, could we set up a key-value to return the translation function, e.g. `encodingValueShape`, depending on the `aesName`?

Sorry if we have tried this already and it doesn't work...